src/State/ForKeys.lua

```diff
@@ -0,0 +1,231 @@
+		-- get or create key data
+		local keyData = self._keyData[newInKey]
+		if keyData == nil then
+			keyData = {
+				-- we don't need strong references here - the main set does that
+				-- for us, so let's not introduce unnecessary leak opportunities
+				dependencySet = setmetatable({}, WEAK_KEYS_METATABLE),
+				oldDependencySet = setmetatable({}, WEAK_KEYS_METATABLE),
+				dependencyValues = setmetatable({}, WEAK_KEYS_METATABLE),
+			}
+			self._keyData[newInKey] = keyData
+		end
+
+		-- if an input key's previous value is non-nil, then there's no need to recalculate it
+		-- this also allows it to have a non-truthy value, which is important since we don't care
+		-- about the value; if we have dependencies then we also need to check if they've changed,
+		-- since if they have then the key has "changed".
+		local shouldRecalculate = oldInputKeys[newInKey] == nil
+
+		if not shouldRecalculate then
+			-- check if dependencies have changed
+			for dependency, oldValue in pairs(keyData.dependencyValues) do
+				-- if the dependency changed value, then this needs recalculating
+				if oldValue ~= dependency:get(false) then
+					shouldRecalculate = true
+					break
+				end
+			end
+		end
+
+		-- if we should recalculate the output by this point, do that
+		if shouldRecalculate then
+			keyData.oldDependencySet, keyData.dependencySet = keyData.dependencySet, keyData.oldDependencySet
+			table.clear(keyData.dependencySet)
+
+			local processOK, newOutKey, newMetaValue = captureDependencies(
+				keyData.dependencySet,
+				self._processor,
+				newInKey
+			)
+
+			if processOK then
+				local oldInKey = keyOIMap[newOutKey]
+
+				-- if there are colliding output keys, throw an error
+				if oldInKey ~= newInKey and newInputKeys[oldInKey] ~= nil then
+					logError("forKeysKeyCollision", nil, tostring(newOutKey), tostring(oldInKey), tostring(newOutKey))
+				end
+
+				-- make the old input match the new input value
+				oldInputKeys[newInKey] = _value
+				-- store the new meta value in the table
+				meta[newOutKey] = newMetaValue
+				-- store the new output key for next time we run the output comparison
+				keyOIMap[newOutKey] = newInKey
+				-- store the new output key in the table with its original value, which we give to the user
+				outputKeys[newOutKey] = _value
+			else
+				-- restore old dependencies, because the new dependencies may be corrupt
+				keyData.oldDependencySet, keyData.dependencySet = keyData.dependencySet, keyData.oldDependencySet
+
+				logErrorNonFatal("forKeysProcessorError", newOutKey)
+			end
+		end
+	end
+
+	-- STEP 2: find keys that were removed
+
+	for outputKey, inputKey in pairs(keyOIMap) do
+		-- if the output key doesn't have an equivalent input key in the new input table
+		if newInputKeys[inputKey] == nil then
+			-- clean up the old calculated value
+			local oldMetaValue = meta[outputKey]
+
+			local destructOK, err = xpcall(self._destructor, parseError, outputKey, oldMetaValue)
```

Comment:
Although only the keys are transformed, may values still need to be cleaned up? They are currently not passed to the destructor.

Reply:
This was discussed in #10. The destructor should clean up things created in the processor, nothing else. Because the value remains unchanged by the processor, it is up to the user to clean them up. We shouldn't make assumptions on when something that we don't explicitly create needs to be cleaned up: what if the user still needs to use the value after the key referencing it is destroyed?

 Besides, if someone did something like this:

```lua
local state = Value({ ["key"] = Instance.new("Part"), })

local computed = ForKeys(state, function(key)
  return key
end)

state:set({})
```

The value would be automatically be cleaned up since it's no longer referenced anywhere.